pkg/boot/uefi/uefi.go

```diff
@@ -0,0 +1,178 @@
+const (
+	SerialPortTypeIO   = 1
+	SerialPortTypeMMIO = 2
+)
+
+// SerialPortConfig defines debug port configuration
+type SerialPortConfig struct {
+	Type        uint32
+	BaseAddr    uint32
+	Baud        uint32
+	RegWidth    uint32
+	InputHertz  uint32
+	UartPciAddr uint32
+}
+
+type payloadConfig struct {
+	AcpiBase            uint64
+	AcpiSize            uint64
+	SmbiosBase          uint64
+	SmbiosSize          uint64
+	SerialConfig        SerialPortConfig
+	NumMemoryMapEntries uint32
+}
+
+// FvImage is a structure for loading a firmware volume
+type FvImage struct {
+	name         string
+	mem          kexec.Memory
+	entryAddress uintptr
+	ImageBase    uintptr
+	SerialConfig SerialPortConfig
+}
+
+// PEImage is a structure for loading entry PE image from a firmware volume
+type PEImage struct {
+	file *os.File
+}
+
+// ReadAt of PEImage will add fvEntryImageOffset to regular ReadAt's offset
+func (fv *PEImage) ReadAt(p []byte, off int64) (n int, err error) {
+	// Entry PE image for UEFI Payload is 0xA0 offset from FV head.
+	return fv.file.ReadAt(p, off+fvEntryImageOffset)
+}
+
+func checkFvAndGetEntryPoint(name string) (uintptr, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	pf, err := pe.NewFile(&PEImage{file: f})
+	if err != nil {
+		return 0, err
+	}
+	return uintptr(fvEntryImageOffset) + uintptr(pf.OptionalHeader.(*pe.OptionalHeader64).AddressOfEntryPoint), nil
+}
+
+// New loads the file and return FvImage stucture if entry image is found
+func New(n string) (*FvImage, error) {
+	entry, err := checkFvAndGetEntryPoint(n)
+	if err != nil {
+		return nil, err
+	}
+	return &FvImage{name: n, mem: kexec.Memory{}, entryAddress: entry}, nil
+}
+
+// Reserved 64kb for passing params
+const uefiPayloadConfigSize = 0x10000
+
+// Load loads fimware volume payload and boot the the payload
+func (fv *FvImage) Load(verbose bool) error {
+	// Install payload
+	dat, err := ioutil.ReadFile(fv.name)
+	if err != nil {
+		return err
+	}
+	fv.mem.Segments.Insert(kexec.NewSegment(dat, kexec.Range{Start: fv.ImageBase, Size: uint(len(dat))}))
+
+	// Install payload config & its memory map: 64 kb below the image
+	// We cannot use the memory above  the image base because it may be used by HOB
+	var configAddr uintptr = fv.ImageBase - uintptr(uefiPayloadConfigSize)
+
+	// Get MemoryMap
+	mm, err := kexecParseMemoryMap()
+	if err != nil {
+		return err
+	}
+
+	// Get Acpi Basc (RSDP)
+	rsdp, err := getRSDP()
+	if err != nil {
+		return err
+	}
+
+	smbiosBase, smbiosSize, err := getSMBIOSBase()
+	if err != nil {
+		return err
+	}
+
+	pc := payloadConfig{
+		AcpiBase:            uint64(rsdp.RSDPAddr()),
+		AcpiSize:            uint64(rsdp.Len()),
+		SmbiosBase:          uint64(smbiosBase),
+		SmbiosSize:          uint64(smbiosSize),
+		SerialConfig:        fv.SerialConfig,
+		NumMemoryMapEntries: uint32(len(mm)),
+	}
+
+	pcbuf := &bytes.Buffer{}
+	if err := binary.Write(pcbuf, binary.LittleEndian, pc); err != nil {
```

Comment:
Should you check the struct's "version" matches? In case you need to add/remove structs in the future?

Reply:
This is the one part I am also struggled. Normally, we add struct version to block it. However, if there are consistent issue between struct, the only think we can do is crash because we already kexec into the binary. To be more robust, I think we may need to version control between payload and u-root version. However, I think we should do it in other place (future work?). WDYT.